# Classification Project: Telco Churn

## Table of Contents

* [Package Imports](#packages_import)
* [Acquire](#data_import)
* [Prepare](#prepare)
* [Explore Process](#explore1)
    * [Question 1](#q_1)
        * [Q1 Visualization](#q_1_viz)
        * [Q1 Statistical Test](#q_1_stats)
        * [Q1 Answer](#q_1_ans)
    * [Question 2](#q_2)
        * [Q2 Visualization](#q_2_viz)
        * [Q2 Statistical Test](#q_2_stats)
        * [Q2 Answer](#q_2_ans)    
    * [Question 3](#q_3)
        * [Q3 Visualization](#q_3_viz)
        * [Q3 Statistical Test](#q_3_stats)
        * [Q3 Answer](#q_3_ans)    
    * [Question 4](#q_4)
        * [Q4 Visualization](#q_4_viz)
        * [Q4 Statistical Test](#q_4_stats)
        * [Q4 Answer](#q_4_ans)
* [Explore Summary](#explore2)
* [Modeling](#modeling)
    * [Introduction](#m_intro)
    * [Baseline](#baseline)
    * [Model 1](#mod_1)
    * [Model 2](#mod_2)
    * [Model 3](#mod_3)
* [Conclusion](#conclusion)
    * [Summery](#c_summery)
    * [Recommendations](#c_recs)
    * [Next Steps](#c_steps)
* [ReadMe](#readme)

## Package Imports <a class="anchor" id="packages_import"></a>


In [1]:
#Importing required packages and files
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
from acquire import get_telco_data
from prepare import prep_telco, train_validate

pd.options.display.max_columns = None

## Acquire<a class="anchor" id="data_import"></a>

In [2]:
#First tries to get telco data from catche or aquires new data from MySQL database and assigns it to the master Data Frame: 
master_df = get_telco_data()
len(master_df[master_df.churn_month == '2022-01-31']) / len(master_df)


0.2653698707936959

## Prepare <a class="anchor" id="prepare"></a>

In [3]:
#Takes the master_df and prepares the data for statistical and machine learning analysis:
working_df = prep_telco(master_df)
working_df.head()

paperless_billing  senior_citizen  partner  dependents  monthly_charges  \
0                  1               0        1           1             65.6   
1                  0               0        0           0             59.9   
2                  1               0        0           0             73.9   
3                  1               1        1           0             98.0   
4                  1               1        1           0             83.9   

   total_charges  churn  tenure  gender_Female  gender_Male  \
0         593.30      0     9.0              1            0   
1         542.40      0     9.0              0            1   
2         280.85      1     4.0              0            1   
3        1237.85      1    13.0              0            1   
4         267.40      1     3.0              1            0   

   streaming_movies_No  streaming_movies_No internet service  \
0                    1                                     0   
1                    0                                     0   
2                    1                                     0   
3                    0                                     0   
4                    1                                     0   

   streaming_movies_Yes  streaming_tv_No  streaming_tv_No internet service  \
0                     0                0                                 0   
1                     1                1                                 0   
2                     0                1                                 0   
3                     1                0                                 0   
4                     0                0                                 0   

   streaming_tv_Yes  tech_support_No  tech_support_No internet service  \
0                 1                0                                 0   
1                 0                1                                 0   
2                 0                1                                 0   
3                 1                1                                 0   
4                 1                0                                 0   

   tech_support_Yes  multiple_lines_No  multiple_lines_No phone service  \
0                 1                  1                                0   
1                 0                  0                                0   
2                 0                  1                                0   
3                 0                  1                                0   
4                 1                  1                                0   

   multiple_lines_Yes  online_backup_No  online_backup_No internet service  \
0                   0                 0                                  0   
1                   1                 1                                  0   
2                   0                 1                                  0   
3                   0                 0                                  0   
4                   0                 1                                  0   

   online_backup_Yes  online_security_No  online_security_No internet service  \
0                  1                   1                                    0   
1                  0                   1                                    0   
2                  0                   1                                    0   
3                  1                   1                                    0   
4                  0                   1                                    0   

   online_security_Yes  device_protection_No  \
0                    0                     1   
1                    0                     1   
2                    0                     0   
3                    0                     0   
4                    0                     1   

   device_protection_No internet service  device_protection_Yes  \
0                                      0                      0   
1                         

In [4]:
#Seperate data into train(35%), validate(35%) and test(30%) Data Frames:
#Random seed of 1969
train_df, val_df, test_df = train_validate(working_df, stratify_col='churn')
sum(train_df.churn) / len(train_df.churn)

0.2653144016227181

In [5]:
train_df.head()

paperless_billing  senior_citizen  partner  dependents  monthly_charges  \
384                   1               0        1           1            43.95   
6320                  0               0        1           1            25.25   
6781                  0               0        1           1            19.70   
3922                  1               0        1           0           103.90   
5775                  1               0        0           0            49.25   

      total_charges  churn  tenure  gender_Female  gender_Male  \
384         2960.10      0    67.0              1            0   
6320        1573.05      0    62.0              0            1   
6781        1275.85      0    65.0              0            1   
3922        2254.20      1    22.0              1            0   
5775         255.60      0     5.0              1            0   

      streaming_movies_No  streaming_movies_No internet service  \
384                     0                                     0   
6320                    0                                     1   
6781                    0                                     1   
3922                    0                                     0   
5775                    0                                     0   

      streaming_movies_Yes  streaming_tv_No  streaming_tv_No internet service  \
384                      1                1                                 0   
6320                     0                0                                 1   
6781                     0                0                                 1   
3922                     1                0                                 0   
5775                     1                0                                 0   

      streaming_tv_Yes  tech_support_No  tech_support_No internet service  \
384                  0                1                                 0   
6320                 0                0                                 1   
6781                 0                0                                 1   
3922                 1                1                                 0   
5775                 1                0                                 0   

      tech_support_Yes  multiple_lines_No  multiple_lines_No phone service  \
384                  0                  0                                1   
6320                 0                  0                                0   
6781                 0                  1                                0   
3922                 0                  0                                0   
5775                 1                  0                                1   

      multiple_lines_Yes  online_backup_No  online_backup_No internet service  \
384                    0                 1                                  0   
6320                   1                 0                                  1   
6781                   0                 0                                  1   
3922                   1                 1                                  0   
5775                   0                 1                                  0   

      online_backup_Yes  online_security_No  \
384                   0                   0   
6320                  0                   0   
6781                  0                   0   
3922                  0                   0   
5775                  0                   1   

      online_security_No internet service  online_security_Yes  \
384                                     0                    1   
6320                                    1                    0   
6781                                    1                    0   
3922                                    0                    1   
5775                                    0                    0   

      device_protection_No  device_protection_No internet service  \
384                      0                                      

## Explore process <a class="anchor" id="explore1"></a>

In [6]:
#Split into a churn and not churn DataFrame to explore:
churn_df = train_df[train_df.churn == 1]
print(f'churn_df length: {len(churn_df)}')
no_churn = train_df[train_df.churn == 0]
print(f'no_churn length: {len(no_churn)}')

churn_df length: 654
no_churn length: 1811


Takeaway: there are 26.5% of people that churned and 73.5% have not churned.

In [38]:
#Printing a comparison between the baseline and two DataFrames:
print('Column ------------------------------------- Churn ----- Deviation ---- Baseline --- Deviation --- No Churn')
for item in churn_df:
    bl = round(sum(train_df[item]) / len(train_df[item]) * 100,1)
    churn = round(sum(churn_df[item]) / len(churn_df[item]) * 100,1)
    no_c = round(sum(no_churn[item]) / len(no_churn[item]) * 100,1)
    
    
    print(f'{item:38} | {churn:10} % | {churn - bl:10.1f} | {bl:10}% | {no_c - bl:10.1f} | {no_c:10} % |')
    print('')

Column ------------------------------------- Churn ----- Deviation ---- Baseline --- Deviation --- No Churn
paperless_billing                      |       75.7 % |       15.0 |       60.7% |       -5.4 |       55.3 % |

senior_citizen                         |       24.5 % |        8.4 |       16.1% |       -3.0 |       13.1 % |

partner                                |       36.7 % |      -12.9 |       49.6% |        4.6 |       54.2 % |

dependents                             |       17.3 % |      -12.5 |       29.8% |        4.5 |       34.3 % |

monthly_charges                        |     7393.9 % |      931.0 |     6462.9% |     -336.2 |     6126.7 % |

total_charges                          |   156445.6 % |   -74413.3 |   230858.9% |    26872.7 |   257731.6 % |

churn                                  |      100.0 % |       73.5 |       26.5% |      -26.5 |        0.0 % |

tenure                                 |     1830.7 % |    -1427.2 |     3257.9% |      515.4 |     3773.3 %

Features with a high or low deviation from the baseline to focus on: 
1. contract_type_Month-to-month (high) 33.6%
- internet_service_type_Fiber optic (high) 25.0%
- payment_type_Electronic check (high) 22.9%
- device_protection_No internet service (low) -15.6%
- online_security_No internet service (low) -15.6%
- online_backup_No internet service (low) -15.6%
- tech_support_No internet service (low) -15.6%
- streaming_tv_No internet service (low) -15.6%
- streaming_movies_No internet service (low) -15.6%
- dependents (low) -12.5%
- paperless_billing (high) 15.0%

In [13]:
churn_df.monthly_charges.describe()

count    654.000000
mean      73.938838
std       24.614680
min       18.850000
25%       55.312500
50%       78.950000
75%       93.850000
max      116.200000
Name: monthly_charges, dtype: float64

In [15]:
no_churn.monthly_charges.describe()

count    1811.000000
mean       61.266759
std        31.084016
min        18.400000
25%        25.150000
50%        64.450000
75%        87.475000
max       118.600000
Name: monthly_charges, dtype: float64

In [14]:
churn_df.total_charges.describe()

count     654.000000
mean     1564.455887
std      1939.229536
min        18.850000
25%       113.912500
50%       658.400000
75%      2413.112500
max      8109.800000
Name: total_charges, dtype: float64

In [16]:
no_churn.total_charges.describe()

count    1811.000000
mean     2577.315737
std      2338.554452
min         0.000000
25%       564.500000
50%      1752.450000
75%      4329.100000
max      8670.100000
Name: total_charges, dtype: float64

In [8]:
#Split, Apply, Combine.
#Split using a single column train_df = train_df.groupby(['column'])
#get the new group new_group.get_group('group')

#Useful bc you don't need a filter every time:
#new_group['column'].value_counts().loc['any_group']
#To see percentages versus numbers:
#new_group['column'].value_counts(normalize=True).loc['any_group']


#Alt way:
#filt = df['column'] == 'group'
#df.loc[filt]

### Question 1 <a class="anchor" id="q_1"></a>

#### Q1 Visualization <a class="anchor" id="q_1_viz"></a>

#### Q1 Statistical Test <a class="anchor" id="q_1_stats"></a>

#### Q1 Answer <a class="anchor" id="q_1_ans"></a>

### Question 2 <a class="anchor" id="q_2"></a>

#### Q2 Visualization <a class="anchor" id="q_2_viz"></a>

#### Q2 Statistical Test <a class="anchor" id="q_2_stats"></a>

#### Q2 Answer <a class="anchor" id="q_2_ans"></a>

### Question 3 <a class="anchor" id="q_3"></a>

#### Q3 Visualization <a class="anchor" id="q_3_viz"></a>

#### Q3 Statistical Test <a class="anchor" id="q_3_stats"></a>

#### Q3 Answer <a class="anchor" id="q_3_ans"></a>

### Question 4 <a class="anchor" id="q_4"></a>

#### Q4 Visualization <a class="anchor" id="q_4_viz"></a>

#### Q4 Statistical Test <a class="anchor" id="q_4_stats"></a>

#### Q4 Answer <a class="anchor" id="q_4_ans"></a>

## Explore summary <a class="anchor" id="explore2"></a>

## Modeling <a class="anchor" id="modeling"></a>

### Introduction <a class="anchor" id="m_intro"></a>

### Baseline <a class="anchor" id="baseline"></a>

### Model 1 <a class="anchor" id="mod_1"></a>

### Model 2 <a class="anchor" id="mod_2"></a>

### Model 3 <a class="anchor" id="mod_3"></a>

## Conclusion <a class="anchor" id="conclusion"></a>

### Summery <a class="anchor" id="c_summery"></a>

### Recommendations <a class="anchor" id="c_recs"></a>

### Next Steps <a class="anchor" id="c_steps"></a>

## ReadMe <a class="anchor" id="readme"></a>

Your README should contain all of the following elements:

* **Title** Gives the name of your project
* **Project Description** Describes what your project is and why it is important 
* **Project Goal** Clearly states what your project sets out to do and how the information gained can be applied to the real world
* **Initial Hypotheses** Initial questions used to focus your project 
* **Project Plan** Guides the reader through the different stages of the pipeline as they relate to your project
* **Data Dictionary** Gives a definition for each of the features used in your report and the units they are measured in, if applicable
* **Steps to Reproduce** Gives instructions for reproducing your work. i.e. Running your notebook on someone else's computer.